src/hooks/useMultiplayerState.ts

```diff
@@ -63,14 +66,98 @@ export function useMultiplayerState(roomId: string) {
 		[],
 	);
 
-	const onMount = useCallback(
-		(app: TldrawApp) => {
-			app.loadRoom(roomId);
-			app.pause();
-			setAppInstance(app);
-		},
-		[roomId],
-	);
+	async function openFromFileSystem(): Promise<null | {
+		fileHandle: FileSystemFileHandle | null;
+		document: TDDocument;
+	}> {
+		const blob = await fileOpen({
+			description: 'Tldraw File',
+			extensions: [`.tldr`],
+			multiple: false,
+		});
+
+		if (!blob) return null;
+
+		const json: string = await new Promise((resolve) => {
+			const reader = new FileReader();
+			reader.onloadend = () => {
+				if (reader.readyState === FileReader.DONE) {
+					resolve(reader.result as string);
+				}
+			};
+			reader.readAsText(blob, 'utf8');
+		});
+
+		const file: TDFile = JSON.parse(json);
+
+		const fileHandle = blob.handle ?? null;
+
+		return {
+			fileHandle,
+			document: file.document,
+		};
+	}
+
+	const updateYMapsInTransaction = (
+		shapes: Record<string, TDShape | undefined>,
+		bindings: Record<string, TDBinding | undefined>,
+		assets: Record<string, TDAsset | undefined>,
+	) => {
+		doc.transact(() => {
+			Object.entries(shapes).forEach(([id, shape]) => {
+				if (!shape) {
+					yShapes.delete(id);
+				} else {
+					yShapes.set(shape.id, shape);
+				}
+			});
+			Object.entries(bindings).forEach(([id, binding]) => {
+				if (!binding) {
+					yBindings.delete(id);
+				} else {
+					yBindings.set(binding.id, binding);
+				}
+			});
+			Object.entries(assets).forEach(([id, asset]) => {
+				if (!asset) {
+					yAssets.delete(id);
+				} else {
+					yAssets.set(asset.id, asset);
+				}
+			});
+		});
+	};
+
+	const onMount = useCallback((app: TldrawApp) => {
+		app.loadRoom(roomId);
+		app.pause();
```

Comment:
This return null should it? It exists no comment that implemantion should add later. Should it work without await?
The same for loadRoom